testdata/go/v3/memcached-operator/monitoring/metricsdocs/metricsdocs.go

```diff
@@ -0,0 +1,115 @@
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"github.com/example/memcached-operator/monitoring"
+	"sort"
+)
+
+// constant parts of the file
+const (
+	title      = "# Memcached Operator Metrics\n"
+	background = "This document aims to help users that are not familiar with metrics exposed by the memcached-operator.\n" +
+		"All metrics documented here are auto-generated by the utility tool `monitoring/metricsdocs` and reflects exactly what is being exposed.\n\n"
+
+	KVSpecificMetrics = "## Memcached Operator Metrics List\n"
+
+	opening = title +
+		background +
+		KVSpecificMetrics
+
+	// footer
+	footerHeading = "## Developing new metrics\n"
+	footerContent = "After developing new metrics or changing old ones, please run `make generate-metricsdocs` to regenerate this document.\n\n" +
+		"If you feel that the new metric doesn't follow these rules, please change `monitoring/metricsdocs` according to your needs.\n"
+
+	footer = footerHeading + footerContent
+)
+
+// MemcachedMetricList contains the name, description, and type for each metric.
+func main() {
+	MemcachedMetricList := metricDescriptionListToMetricList(monitoring.ListMetrics())
+	sort.Sort(MemcachedMetricList)
+	writeToFile(MemcachedMetricList)
+}
+
+// writeToFile receives a list of metrics and writes them to a file.
+func writeToFile(metricsList metricList) {
+	fmt.Print(opening)
+	metricsList.writeOut()
+	fmt.Print(footer)
+}
+
+// metric is an exported struct that defines the metric
+// name, description, and type as a new type named metric.
+type metric struct {
+	name        string
+	description string
+	metricType  string
+}
+
+// metricDescriptionToMetric receives a metric of type Metric defined
+// in ./monitoring/metrics/util/util.go, and returns as a metric type.
+func metricDescriptionToMetric(md monitoring.MetricDescription) metric {
+	return metric{
+		name:        md.Name,
+		description: md.Help,
+		metricType:  md.Type,
+	}
+}
+
+// writeOut receives a metric of type metric and prints
+// the metric name, description, and type.
+func (m metric) writeOut() {
+	fmt.Println("###", m.name)
+	fmt.Println(m.description, "Type: "+m.metricType+".")
+}
+
+// metricList is an array that contain metrics from type metric,
+// as a new type named metricList.
+type metricList []metric
+
+// metricDescriptionListToMetricList collects the metrics exposed by the
+// memcached-operator, and inserts them into the metricList array.
+func metricDescriptionListToMetricList(mdl []monitoring.MetricDescription) metricList {
+	res := make([]metric, len(mdl))
+	for i, md := range mdl {
+		res[i] = metricDescriptionToMetric(md)
+	}
+
+	return res
+}
+
+// Len implements sort.Interface.Len
+func (m metricList) Len() int {
+	return len(m)
+}
+
+// Less implements sort.Interface.Less
+func (m metricList) Less(i, j int) bool {
+	return m[i].name < m[j].name
+}
+
+// Swap implements sort.Interface.Swap
+func (m metricList) Swap(i, j int) {
+	m[i], m[j] = m[j], m[i]
+}
```

Comment:
@avlitman Where do you use this function?

Reply:
all this implement methods are used for sort: https://github.com/operator-framework/operator-sdk/pull/6114/files#diff-29567a3dab531ff74b6e68ff734786bb53df8bee629bc1d8eaa33809db03105fR45

without using them we will get the error:
monitoring/metricsdocs/metricsdocs.go:45:12: cannot use MemcachedMetricList (variable of type metricList) as type sort.Interface in argument to sort.Sort:
        metricList does not implement sort.Interface (missing **Len/Less/Swap/** method)

Reply:
I tried without sorting, for 1 metric the output is the same, and for 2 or more the A-Z sorting is needed.

Reply:
@sradco @machadovilaca can I resolve this conversation?